File: packages/react-native-nitro-filetoolkit/android/src/main/java/com/margelo/nitro/filetoolkit/FileLocationResolver.kt
package com.margelo.nitro.filetoolkit

import android.content.Context
import android.net.Uri
import com.margelo.nitro.NitroModules
import java.io.File

internal class FileLocationResolver(
  private val context: Context = requireNotNull(NitroModules.applicationContext) {
    "NitroModules has not been installed"
  },
) {
  fun location(directory: ManagedDirectory, relativePath: String): FileLocation {
    validateRelativePath(relativePath)
    val root = root(directory)
    val candidate = File(root, relativePath).canonicalFile
    if (candidate.parentFile == null || !candidate.path.startsWith(root.path + File.separator)) {
      throw FileToolkitException.invalidLocation("relativePath escapes its managed directory")
    }
    return location(candidate, FileLocationOrigin.MANAGED)
  }

  fun fromUri(uri: String): FileLocation = location(fileFromUri(uri), FileLocationOrigin.URI)

  fun file(location: FileLocation): File {
    val file = fileFromUri(location.uri)
    if (location.origin == FileLocationOrigin.MANAGED && managedRoots().none { contains(it, file) }) {
      throw FileToolkitException.invalidLocation("managed location is outside all managed directories")
    }
    return file
  }

  fun root(directory: ManagedDirectory): File {
    val root = rawRoot(directory).canonicalFile
    if ((!root.exists() && !root.mkdirs()) || !root.isDirectory) {
      throw FileToolkitException.invalidOperation("managed directory is unavailable")
    }
    return root
  }

  fun rootLocation(directory: ManagedDirectory): FileLocation =
    location(root(directory), FileLocationOrigin.MANAGED)

  private fun rawRoot(directory: ManagedDirectory): File = when (directory) {
    ManagedDirectory.CACHE -> context.cacheDir
    ManagedDirectory.DOCUMENTS -> context.filesDir
    ManagedDirectory.DOWNLOADS -> File(context.filesDir, "Downloads")
    ManagedDirectory.TEMPORARY -> File(context.cacheDir, "Temporary")
    ManagedDirectory.APPLICATION_SUPPORT -> File(context.filesDir, "ApplicationSupport")
  }.absoluteFile

  private fun fileFromUri(value: String): File {
    val uri = try {
      Uri.parse(value)
    } catch (error: Exception) {
      throw FileToolkitException.invalidLocation("URI is malformed", error)
    }
    if (uri.scheme != "file" || !uri.isAbsolute || !uri.authority.isNullOrEmpty()) {
      throw FileToolkitException.invalidLocation("only absolute file:// URIs are accepted")
    }
    val path = uri.path
    if (path == null || !path.startsWith('/')) {
      throw FileToolkitException.invalidLocation("file URI must contain an absolute path")
    }
    return try {
      File(path).canonicalFile
    } catch (error: Exception) {
      throw FileToolkitException.invalidLocation("file URI is invalid", error)
    }
  }

  private fun managedRoots(): Array<File> = arrayOf(
    ManagedDirectory.CACHE,
    ManagedDirectory.DOCUMENTS,
    ManagedDirectory.DOWNLOADS,
    ManagedDirectory.TEMPORARY,
    ManagedDirectory.APPLICATION_SUPPORT,
  ).map { rawRoot(it).canonicalFile }.toTypedArray()

  private fun contains(root: File, candidate: File): Boolean =
    candidate == root || candidate.path.startsWith(root.path + File.separator)

  private fun location(file: File, origin: FileLocationOrigin): FileLocation {
    val canonical = file.canonicalFile
    return FileLocation(origin, Uri.fromFile(canonical.absoluteFile).toString())
  }

  private fun validateRelativePath(path: String) {
    if (path.isEmpty()) {
      throw FileToolkitException.invalidLocation("relativePath must not be empty")
    }
    if (path.startsWith('/') || path.contains('\\') || path.contains('\u0000')) {
      throw FileToolkitException.invalidLocation("relativePath must be a portable relative path")
    }
    if (path.split('/').any { it.isEmpty() || it == "." || it == ".." }) {
      throw FileToolkitException.invalidLocation(
        "relativePath cannot contain empty, '.' or '..' segments",
      )
    }
  }
}

File: packages/react-native-nitro-filetoolkit/ios/FileLocationResolver.swift
import Foundation

final class FileLocationResolver {
  private let fileManager: FileManager

  init(fileManager: FileManager = .default) {
    self.fileManager = fileManager
  }

  func location(directory: ManagedDirectory, relativePath: String) throws -> FileLocation {
    guard !relativePath.isEmpty else {
      throw FileToolkitError.invalidLocation("relativePath must not be empty")
    }
    guard !relativePath.hasPrefix("/"), !relativePath.contains("\\"), !relativePath.contains("\0") else {
      throw FileToolkitError.invalidLocation("relativePath must be a portable relative path")
    }

    let components = relativePath.split(separator: "/", omittingEmptySubsequences: false)
    guard components.allSatisfy({ !$0.isEmpty && $0 != "." && $0 != ".." }) else {
      throw FileToolkitError.invalidLocation("relativePath cannot contain empty, '.' or '..' segments")
    }

    let root = try rootURL(for: directory).standardizedFileURL
    let candidate = root.appendingPathComponent(relativePath, isDirectory: false).standardizedFileURL
    guard candidate.path.hasPrefix(root.path + "/") else {
      throw FileToolkitError.invalidLocation("relativePath escapes its managed directory")
    }
    return FileLocation(uri: candidate.absoluteString)
  }

  func url(from location: FileLocation) throws -> URL {
    try url(fromUri: location.uri)
  }

  func fromUri(_ uri: String) throws -> FileLocation {
    FileLocation(uri: try url(fromUri: uri).absoluteString)
  }

  func rootURL(for directory: ManagedDirectory) throws -> URL {
    switch directory {
    case .cache:
      return try requiredURL(for: .cachesDirectory)
    case .documents:
      return try requiredURL(for: .documentDirectory)
    case .downloads:
      return try requiredURL(for: .documentDirectory).appendingPathComponent("Downloads", isDirectory: true)
    case .temporary:
      return fileManager.temporaryDirectory
    case .applicationSupport:
      return try requiredURL(for: .applicationSupportDirectory)
    }
  }

  private func url(fromUri uri: String) throws -> URL {
    guard let url = URL(string: uri), url.isFileURL, url.scheme == "file" else {
      throw FileToolkitError.invalidLocation("only absolute file:// URIs are accepted")
    }
    guard url.path.hasPrefix("/") else {
      throw FileToolkitError.invalidLocation("file URI must contain an absolute path")
    }
    return url.standardizedFileURL
  }

  private func requiredURL(for directory: FileManager.SearchPathDirectory) throws -> URL {
    guard let url = fileManager.urls(for: directory, in: .userDomainMask).first else {
      throw FileToolkitError.invalidOperation("managed directory is unavailable")
    }
    return url
  }
}
